File: kernel/common/src/process/scheduler.c
// NebulaOS - Round-Robin Scheduler
// ==================================
//
// Simple round-robin process scheduler

#include "../../common/include/scheduler.h"
#include "../../common/include/process.h"
#include "../../common/include/nebula.h"
#include "../../common/include/stdint.h"

static uint32_t scheduler_queue[MAX_PROCESSES];
static int queue_head = 0;
static int queue_tail = 0;
static int queue_count = 0;
static uint32_t current_scheduled_pid = 0;
static uint32_t tick_counter = 0;
static bool scheduler_initialized = false;

void scheduler_init(void) {
    for (int i = 0; i < MAX_PROCESSES; i++) {
        scheduler_queue[i] = 0;
    }
    queue_head = 0;
    queue_tail = 0;
    queue_count = 0;
    current_scheduled_pid = 0;
    tick_counter = 0;
    scheduler_initialized = true;
}

void scheduler_add(uint32_t pid) {
    if (!scheduler_initialized) return;
    if (queue_count >= MAX_PROCESSES) return;
    
    scheduler_queue[queue_tail] = pid;
    queue_tail = (queue_tail + 1) % MAX_PROCESSES;
    queue_count++;
}

void scheduler_remove(uint32_t pid) {
    if (!scheduler_initialized) return;
    
    for (int i = 0; i < queue_count; i++) {
        int idx = (queue_head + i) % MAX_PROCESSES;
        if (scheduler_queue[idx] == pid) {
            for (int j = i; j < queue_count - 1; j++) {
                int cur = (queue_head + j) % MAX_PROCESSES;
                int nxt = (queue_head + j + 1) % MAX_PROCESSES;
                scheduler_queue[cur] = scheduler_queue[nxt];
            }
            queue_tail = (queue_tail - 1 + MAX_PROCESSES) % MAX_PROCESSES;
            queue_count--;
            return;
        }
    }
}

void scheduler_tick(void) {
    if (!scheduler_initialized || queue_count == 0) return;
    
    tick_counter++;
    if (tick_counter < SCHEDULER_TIME_SLICE) return;
    
    tick_counter = 0;
    
    uint32_t old_pid = scheduler_queue[queue_head];
    queue_head = (queue_head + 1) % MAX_PROCESSES;
    queue_tail = (queue_tail + 1) % MAX_PROCESSES;
    scheduler_queue[queue_tail] = old_pid;
    
    current_scheduled_pid = scheduler_queue[queue_head];
    process_set_current(current_scheduled_pid);
    process_t* cur = process_get_current();
    if (cur) cur->state = PROC_RUNNING;
}

File: kernel/src/common/scheduler.rs
use crate::common::stdint::*;
use crate::common::process::{self, Process, PROC_READY};

pub const SCHEDULER_TIME_SLICE: u32 = 10;

static mut SCHEDULER_QUEUE: [u32; crate::common::process::MAX_PROCESSES] = [0; crate::common::process::MAX_PROCESSES];
static mut QUEUE_HEAD: usize = 0;
static mut QUEUE_TAIL: usize = 0;
static mut QUEUE_COUNT: usize = 0;
static mut CURRENT_SCHEDULED_PID: u32 = 0;
static mut TICK_COUNTER: u32 = 0;

pub unsafe fn scheduler_init() {
    for i in 0..crate::common::process::MAX_PROCESSES {
        SCHEDULER_QUEUE[i] = 0;
    }
    QUEUE_HEAD = 0;
    QUEUE_TAIL = 0;
    QUEUE_COUNT = 0;
    CURRENT_SCHEDULED_PID = 0;
    TICK_COUNTER = 0;
}

pub unsafe fn scheduler_add(pid: u32) {
    if QUEUE_COUNT >= crate::common::process::MAX_PROCESSES {
        return;
    }
    SCHEDULER_QUEUE[QUEUE_TAIL] = pid;
    QUEUE_TAIL = (QUEUE_TAIL + 1) % crate::common::process::MAX_PROCESSES;
    QUEUE_COUNT += 1;
}

pub unsafe fn scheduler_remove(pid: u32) {
    for i in 0..QUEUE_COUNT {
        let idx = (QUEUE_HEAD + i) % crate::common::process::MAX_PROCESSES;
        if SCHEDULER_QUEUE[idx] == pid {
            for j in i..QUEUE_COUNT - 1 {
                let cur = (QUEUE_HEAD + j) % crate::common::process::MAX_PROCESSES;
                let nxt = (QUEUE_HEAD + j + 1) % crate::common::process::MAX_PROCESSES;
                SCHEDULER_QUEUE[cur] = SCHEDULER_QUEUE[nxt];
            }
            QUEUE_TAIL = (QUEUE_TAIL - 1 + crate::common::process::MAX_PROCESSES) % crate::common::process::MAX_PROCESSES;
            QUEUE_COUNT -= 1;
            return;
        }
    }
}

pub unsafe fn scheduler_tick() {
    if QUEUE_COUNT == 0 {
        return;
    }
    TICK_COUNTER += 1;
    if TICK_COUNTER < SCHEDULER_TIME_SLICE {
        return;
    }
    TICK_COUNTER = 0;
    let old_pid = SCHEDULER_QUEUE[QUEUE_HEAD];
    QUEUE_HEAD = (QUEUE_HEAD + 1) % crate::common::process::MAX_PROCESSES;
    QUEUE_TAIL = (QUEUE_TAIL + 1) % crate::common::process::MAX_PROCESSES;
    SCHEDULER_QUEUE[QUEUE_TAIL] = old_pid;
    CURRENT_SCHEDULED_PID = SCHEDULER_QUEUE[QUEUE_HEAD];
    process::process_set_current(CURRENT_SCHEDULED_PID);
    if let Some(cur) = (CURRENT_SCHEDULED_PID as usize).checked_sub(1).and_then(|idx| PROCESS_TABLE.get_mut(idx)) {
        cur.state = PROC_RUNNING;
    }
}
